### PR TITLE
[WIP] Adding LID (Language Identification) Head for M-CTC-T Model

### PR DESCRIPTION
**Pull request by cwkeam**
# What does this PR do?

Adding LID head to the M-CTC-T Model.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by cwkeam** [comment]
Developing this as a `SequenceClassification` head. This is correct right? @patrickvonplaten 

TODO:
- [ ] Export LID head weights from flashlight model
- [ ] make new `cwkeam/m-ctct-t-large-lid` checkpoint with above exported weights
- [ ] write tests

**Comment by patrickvonplaten** [comment]
This looks good to me @lorenlugosch and @anton-l  what do you think?

**Comment by lorenlugosch** [comment]
Looks good---I'll give you the mapping from logit index to language shortly.

File: src/transformers/models/mctct/modeling_mctct.py [hunk 1]
```diff
@@ -823,3 +828,105 @@ def forward(
...
+    Language Identification (LID).
+    """,
+    MCTCT_START_DOCSTRING,
+)
+class MCTCTForSequenceClassification(MCTCTPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.mctct = MCTCTModel(config)
+        num_layers = config.num_hidden_layers + 1  # transformer layers + input embeddings
```
Comment (anton-l) [inline src/transformers/models/mctct/modeling_mctct.py:845]:
```suggestion
```

**Review by anton-l — COMMENTED** [review]
Yeah, this should be perfect for LID, thanks @cwkeam!

**Review by lorenlugosch — COMMENTED** [review]
Could add an option for users to decide? 
```
if reduce_mean: return logits.mean(dim=1)
else: return logits
```

**Comment by anton-l** [comment]
@lorenlugosch if that was the intention for the original implementation then yes, for sure! We can also turn the model into `MCTCForAudioFrameClassification` (similar to `Wav2Vec2ForAudioFrameClassification`) to make it clearer that the outputs are frame-level.

**Comment by cwkeam** [comment]
@anton-l That sounds like the perfect solution! Will get going with it.

**Comment by patrickvonplaten** [comment]
@anton-l do you want to help finalize the PR here or should I do it?

**Comment by cwkeam** [comment]
@patrickvonplaten @anton-l 

Firstly, I apologize for being irresponsible with going through with something I started. Life really got in the way here but this is so close to being done.

I've put up the weights as well from my last run at this: https://huggingface.co/cwkeam/m-ctc-t-large-lid

**Comment by anton-l** [comment]
@cwkeam do you need a hand with any part that's left (tests, modelcard, etc.)? I'd be happy to help! :hugs:

**Comment by cwkeam** [comment]
Actually, I realized I actually had finished the full working implementation. I think because of little nits I didn't fully notify you guys but I think it's worth a review:

```python
from transformers import MCTCTProcessor, MCTCTForAudioFrameClassification
from datasets import load_dataset
import torch

dataset = load_dataset("hf-internal-testing/librispeech_asr_demo", "clean", split="validation")
dataset = dataset.sort("id")
sampling_rate = dataset.features["audio"].sampling_rate

processor = MCTCTProcessor.from_pretrained("cwkeam/m-ctc-t-large-lid")
model = MCTCTForAudioFrameClassification.from_pretrained("cwkeam/m-ctc-t-large-lid")

# audio file is decoded on the fly
inputs = processor(dataset[0]["audio"]["array"], sampling_rate=sampling_rate, return_tensors="pt")
with torch.no_grad():
    logits = model(**inputs).logits
predicted_ids = torch.argmax(logits, dim=-1)
predicted_ids
>>> tensor([[12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,
         12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12]])
```


Where, as you can see from @lorenlugosch 's comment above, 12 is English, which we expect from librispeech demo.

**Comment by anton-l** [comment]
@cwkeam perfect! Then I'll go over the PR this week and prepare it for merging :)

**Comment by patrickvonplaten** [comment]
@anton-l and I won't have any time soon to look into this. Should we let it stale for now?
Or @cwkeam, @lorenlugosch would you be interested in taking this PR over?

**Pull request closed without merging**